### DISCUSSION TITLE
Make CollapsingHeader truncateable

### DISCUSSION THREAD

**Discussion by higumachan** [emilk/egui#4093] [Ideas]
I am currently using a patch that allows the CollapsingHeader to truncate the header. I am considering creating a PR if this patch would be useful to others. What do you think? The API I created for truncation looks like this.

```rust
CollapsingHeader::new("collapsing_header_label")
    .header_truncate()
    .body(|ui| { ... })
```

**Comment by YgorSouza** [comment]
You can make a PR and see if it gets accepted, but know that the similar PR #4045 was rejected. It is relatively easy to make a collapsing header with any UI you want, as shown in this example from the Misc Demos section in the demo app:

https://github.com/emilk/egui/blob/5cf99c6308e36b5f36861accea6835f190924cce/crates/egui_demo_lib/src/demo/misc_demo_window.rs#L439-L450

So arbitrarily adding some of these options to the default header seems redundant.

**Reply by higumachan** [reply]
Thank you for response.

I know that with CollapsingState, I can create an arbitrary widget. However, the problem in this case is that the body does not toggle when the user clicks on the arbitrary widget.

Is there a simple solution to this problem?

Or would it make more sense to create a mode in CollapsingState that toggles from a click on the header? (I have tried to implement that as a PoC as well.)

**Reply by YgorSouza** [reply]
I don't know if there is an easy way to do that currently, because the header response does not sense clicks, and the UI you put inside the header doesn't have to return a `Response`, so the header can't use it. And also the `show_header` method consumes the `CollapsingState`, so you can't toggle it manually afterward. So I guess you would have to store the state yourself using `ui.data`, toggle it if your custom UI's response was clicked, and then force it manually using `set_open` (i.e., duplicate some of the CollapsingState's inner logic).

I think this could be implemented in egui with the following code after this part:

https://github.com/emilk/egui/blob/5cf99c6308e36b5f36861accea6835f190924cce/crates/egui/src/containers/collapsing_header.rs#L147-L153

```rs
let clicked = ui
    .allocate_rect(header_response.response.rect, Sense::click())
    .clicked();
if clicked {
    self.toggle(ui);
}
```

But it has to be opt-in, since it makes the custom UI you put in the header non-clickable. So there would have to be a method like `toggle_if_clicked_on_header` or something like that, which makes this code run.

So maybe you could try pitching that as a feature request issue. I think people are more likely to notice those than the discussions.
